src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs

```diff
@@ -984,7 +984,59 @@ private BoundExpression VisitRangePatternIndexerAccess(BoundImplicitIndexerAcces
             AddPlaceholderReplacement(node.ArgumentPlaceholders[1], rangeSizeExpr);
 
             var sliceCall = (BoundCall)node.IndexerOrSliceAccess;
-            var rewrittenIndexerAccess = VisitExpression(sliceCall);
+
+            // [start..] can be simplified to .Substring(start) or .Slice(start) for some built-in types
+            // e.g. string and (ReadOnly)Span
+
+            BoundExpression? moreEfficientIndexerAccess = null;
+            if (endMakeOffsetInput is null)
+            {
+                MethodSymbol? singleArgumentOverload = null;
+                NamedTypeSymbol? typeContainingElementType = null;
+
+                if (
+                    TryGetSpecialTypeMethod(node.Syntax, SpecialMember.System_String__Substring, out var stringSubstring, isOptional: true)
+                    && sliceCall.Method.Equals(stringSubstring)
+                )
+                {
+                    // If the overload is missing, we won't try the other `else if`
+                    if (TryGetSpecialTypeMethod(node.Syntax, SpecialMember.System_String__SubstringInt, out var stringOverload, isOptional: true))
+                        singleArgumentOverload = stringOverload;
+                }
+                // with single type parameter ((ReadOnly)Span)
+                else if (sliceCall is { Method: SubstitutedMethodSymbol { UnderlyingMethod: var generalizedMethod }, Type: NamedTypeSymbol typeWithElementType })
+                {
+                    // We use the return type of the current overload to simplify the logic
+                    typeContainingElementType = typeWithElementType;
```

Comment:
This is never used in the other path. Consider moving it directly into this `else if` for simplicity and just directly setting `singleArgumentOverload` to the correct result.

Reply:
Managed to fix it